### PR TITLE
handheld-daemon: init at 1.1.0

### PR DESCRIPTION
**Pull request by appsforartists**
## Description of changes

A derivation to enable https://github.com/hhd-dev/hhd to be installed on NixOS.

It is a library that provides hardware support for gaming handhelds like the Lenovo Legion Go.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages

_(no tests exist yet for this package)_

- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by appsforartists** [comment]
cc @antheas @IogaMaster

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (antheas) [inline pkgs/by-name/ha/handheld-daemon/package.nix:44]: Make sure the output path does not need /usr, because on arch and fedora it does.

Comment (antheas) [reply pkgs/by-name/ha/handheld-daemon/package.nix:44]: Not all udev rules are meant to be copied for install. 

Only the non-user variant should if `hhd` runs as a system service, otherwise you introduce a tiny vulnerability and expand the scope of permissions.

**Comment by antheas** [comment]
Reference the arch build script for how to bundle:
https://github.com/hhd-dev/hhd/blob/master/PKGBUILD

I would suggest referencing how nix packages pypi and bundling from that instead of the git. 

It will be easier to maintain and, since pypi is immutable, it is also more secure.

**Comment by antheas** [comment]
The controller issues you are describing stem from both the `hid_playstation` driver not being loaded and your system not having the proper udev rules. 

I think that is due to the path issue but maybe `/usr` is built in to the output path.

In any case, you need to reload the rules or reboot afterwards.
```bash
sudo udevadm control --reload-rules && sudo udevadm trigger
```

I would suggest bundling the playstation driver with nix if you include hhd, as that fixes both the touchpad of the controller in desktop mode and the mappings for the apps that do not support Dualsense natively (wine, browsers).

Note that the kernel input driver (which is loaded when `hid_playstation` is not available) handles Dualsense 5 correctly. But not Dualsense Edge.

**Comment by appsforartists** [comment]
Bundling from PyPi is certainly an option.  I chose git because it makes it easier to test new versions before they land in PyPi.  It's pinned to a commit and verified with a hash, so I don't think it adds any vulnerabilities.

**Comment by appsforartists** [comment]
I have updated the derivation to only copy the file in the `PKGBUILD` you linked:

```nix
  postInstall = ''
    mkdir -p $out/lib/udev/
    install -m644 $src/usr/lib/udev/rules.d/83-hhd.rules $out/lib/udev/rules.d
  '';
```

If you search this repo, you'll see plenty of `$out/lib/udev/rules.d` and no `$out/usr/lib/udev/rules.d`, so I think this is correct.  However, I'm still seeing https://github.com/hhd-dev/hhd/issues/8.

```bash
$ lsmod | grep playstation
hid_playstation        36864  0
led_class_multicolor    16384  1 hid_playstation
ff_memless             16384  2 xpad,hid_playstation
hid                   172032  15 i2c_hid,usbhid,hid_multitouch,hid_sensor_hub,hid_playstation,hid_generic,amd_sfh,uhid
led_class              24576  8 snd_hda_codec_generic,mt76,sdhci,input_leds,xpad,hid_playstation,ideapad_laptop,led_class_multicolor
```

**Comment by appsforartists** [comment]
(I don't know why that random `hugo` commit is showing up in this PR.  I'll rebase to whatever `master` is at when this is ready to land.)

**Comment by antheas** [comment]
Hmmmm, very peculiar. You can use the config to switch to uinput mode. And that will work sans dual sense features

Check dmesg to verify that the controllers are loaded properly by the driver. When the driver is properly loaded this fixes the mappings on games that do not support the hid api (and steam without hidraw access)

Then check that you can see led settings under controller calibration and l4/l5 below the controller in the tester. This means that steam can load the hidraw node of the controller.


I have had the issue you're describing on apps that do not load the hidraw node when the playstation driver is not loaded, because the original Linux driver does not map the edge properly. But maybe it's a new steam bug we need to resolve.

File: pkgs/top-level/all-packages.nix
Comment (antheas) [inline pkgs/top-level/all-packages.nix:3938]: `hhd` or `handheld-daemon`? I'm sure hhd is available.

**Comment by antheas** [comment]
Everyone else does a copy with a specific filename, maybe this is a problem. Example.
`install -Dm755 90-micro-3d-local.rules $out/lib/udev/rules.d/90-micro-3d-local.rules`

**Comment by antheas** [comment]
I have replicated the issue when the playstation driver is loaded but the udev rules are invalid. So the playstation driver is irrelevant. The udev rules do not install properly.

File: maintainers/maintainer-list.nix
Comment (Janik-Haag) [inline maintainers/maintainer-list.nix:1367]: please also add some way of contacting you, so email or matrix.

File: pkgs/top-level/all-packages.nix
Comment (appsforartists) [reply pkgs/top-level/all-packages.nix:3938]: I wasn't sure which to choose, so I went with the more human-readable.

`services.handheldDaemon.enable` gives you a better idea of what you're doing than `services.hhd.enable`.

File: maintainers/maintainer-list.nix
Comment (appsforartists) [reply maintainers/maintainer-list.nix:1367]: I check GitHub more reliably than I check my email. 🙃 

The comment at the top of `maintainers-list.nix` says `githubId` is an acceptable form of communication.  Is that out of date?

Comment (Janik-Haag) [reply maintainers/maintainer-list.nix:1367]: oh wait maybe I misremembered :see_no_evil:

**Comment by appsforartists** [comment]
Sorry - the holidays have been busy, so I'm replying from my phone, but I did get a quick chance to look at this again:

The emulated Edge appeared as `hidraw9` in `dmesg`.  `chmod`ding it did indeed fix the issue temporarily, but I'm not sure how to fix it permanently.  I tried copying over the `modules-load.d` rule as well, but that didn't help.

The `systemd` rule appears to be working, since `hhd` runs when I toggle the setting. I'll double check the `udev` rules next time I'm on the Go. It's building fine, so I expect it's doing what it's supposed to. 

Maybe I need to recruit someone who knows Nix better than I do to take a look.

**Comment by antheas** [comment]
The udev rules are not installed properly. Also mind that you have to reload them or reboot for them to work.

The udev rules do two things now for hhd: fix the hidraw issue and disable the accelerometer for screen orientation (which causes issues).

What you are experiencing is a symptom of the rules not being installed.

**Review by IogaMaster — APPROVED** [review]

**Comment by IogaMaster** [comment]
Other than the issues already posed it looks good!

**Comment by appsforartists** [comment]
Thanks to a tip from @mweinelt, this is working now.

**Comment by appsforartists** [comment]
I rolled back to 0.2.7 due to https://github.com/hhd-dev/hhd/issues/12.

**Comment by antheas** [comment]
I did not notice you did not make a template service file but since you mention it.

Switch to a template service file so the user can specify the username `hhd` should bind to for the configuration files etc. Otherwise `hhd` will not work correctly.

**Comment by appsforartists** [comment]
You mean [this](https://github.com/NixOS/nixpkgs/pull/277661/files#diff-f464e6d989b2b7d3440aed1647719963646e20cbadd10466b7ccf7ecd5d2f411R19)?  It's been there since the beginning, but I guess the `%i` isn't working.

I can add a config value to pass in `userName`.

**Comment by antheas** [comment]
misses setuptools as a run dependency. setuptools is not just required for build, it is also required for plugin discoverability so hhd errors out if the user does not have it in their nix system.

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (h7x4) [inline pkgs/by-name/ha/handheld-daemon/package.nix:43]:
```suggestion
    install -Dm644 $src/usr/lib/udev/rules.d/83-hhd.rules -t $out/lib/udev/rules.d/
```

File: nixos/modules/services/hardware/handheld-daemon.nix
Comment (h7x4) [inline nixos/modules/services/hardware/handheld-daemon.nix:34]: Maybe a package option could be added?

```suggestion
    enable = mkEnableOption "Enable Handheld Daemon";
    
    package = mkPackageOption pkgs "handheld-daemon" { };

    user = mkOption {
      type = types.str;
      description = lib.mdDoc ''
        The user to run Handheld Daemon with.
      '';
    };
  };

  config = mkIf cfg.enable {
    environment.systemPackages = [ pkgs.handheld-daemon ];
    services.udev.packages = [ pkgs.handheld-daemon ];
    systemd.packages = [ pkgs.handheld-daemon ];

    systemd.services.handheldDaemon = {
      description = "Handheld Daemon";

      wantedBy = [ "multi-user.target" ];

      restartIfChanged = true;

      serviceConfig = {
        ExecStart = "${ cfg.package }/bin/hhd --user ${ cfg.user }";
```

File: nixos/modules/services/hardware/handheld-daemon.nix
Comment (h7x4) [inline nixos/modules/services/hardware/handheld-daemon.nix:14]: Would it make sense to add some kind of default?

Comment (antheas) [reply nixos/modules/services/hardware/handheld-daemon.nix:14]: should default to the user of the system if possible

File: nixos/modules/services/hardware/handheld-daemon.nix
Comment (appsforartists) [reply nixos/modules/services/hardware/handheld-daemon.nix:34]: What would that do?

Are you trying to make the rule reusable with other packages?

(This is my first nixpkg.)

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (appsforartists) [reply pkgs/by-name/ha/handheld-daemon/package.nix:43]: I had wondered if there was a better way to do that. Thank you!

File: nixos/modules/services/hardware/handheld-daemon.nix
Comment (h7x4) [reply nixos/modules/services/hardware/handheld-daemon.nix:34]: It would make it easier to override the package, in case you'd want to do any changes to the package. Example:

```nix
services.handheldDaemon.package = pkgs.handheld-daemon.override {
  python3 = pkgs.python38;
};
```

Comment (appsforartists) [reply nixos/modules/services/hardware/handheld-daemon.nix:34]: Interesting.  Wouldn't have expected someone would want to do that.

I'm guessing you'd also need to use `cfg.package` in these places?

```nix
    environment.systemPackages = [ pkgs.handheld-daemon ];
    services.udev.packages = [ pkgs.handheld-daemon ];
    systemd.packages = [ pkgs.handheld-daemon ];
```

File: nixos/modules/services/hardware/handheld-daemon.nix
Comment (appsforartists) [reply nixos/modules/services/hardware/handheld-daemon.nix:14]: I don't know that there's a good default.  If there's no way to get the current user, forcing people to specify seems better than making a guess (like Steam Deck's `deck`) that could be wrong for many users.

**Comment by appsforartists** [comment]
I had previously understood that services were meant to be named camelCased, but the ones on my system are kebab-cased, so I just updated this PR to match.

**Comment by antheas** [comment]
Some of the newer additions to hhd (ROG Ally + GPD Win Mini) require access to `modprobe`

This is the error received when running as the nix service `[Errno 2] No such file or directory: 'modprobe'`

File: nixos/modules/services/hardware/handheld-daemon.nix
Comment (h7x4) [reply nixos/modules/services/hardware/handheld-daemon.nix]: > I'm guessing you'd also need to use `cfg.package` in these places?

Correct

**Comment by appsforartists** [comment]
TODO when we update to 1.1.3:

Set `HHD_MOUNT_TRIG_SYSFS` to `$out/config`, and make that dir.

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (lukegb) [reply pkgs/by-name/ha/handheld-daemon/package.nix:11]: That said feel free to introduce this at 1.1.0 and just update it in a followup PR, and feel free to request a review from me explicitly :)

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (antheas) [reply pkgs/by-name/ha/handheld-daemon/package.nix:30]: It needs to. It is not just used for build.

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (appsforartists) [reply pkgs/by-name/ha/handheld-daemon/package.nix:11]: Thanks for the review!

Let's merge 1.1.0, and then follow up with a version bump.  There are some new environ vars that need to be set in the new version.

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (antheas) [reply pkgs/by-name/ha/handheld-daemon/package.nix:17]: I don't know if it is helpful context, but hhd is versioned and also provided as a source tar ball from pypi.

https://pypi.org/project/hhd/

This nix pr builds from GitHub, but copr and aur both build from pypi.

So it may be better without the v

File: nixos/modules/services/hardware/handheld-daemon.nix
Comment (ambroisie) [inline nixos/modules/services/hardware/handheld-daemon.nix:35]:
```suggestion
        ExecStart = "${lib.getExe cfg.package} --user ${cfg.user}";
```

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (lukegb) [reply pkgs/by-name/ha/handheld-daemon/package.nix:17]: In general I prefer builds from GitHub; that way it's more obvious that the source is what you're building. Pypi tarballs are provided by the uploader, so can be quite different from the source on GitHub...

Comment (ambroisie) [reply pkgs/by-name/ha/handheld-daemon/package.nix:17]: It's better to use `rev = "v${version}" (or rev = "refs/tags/v${version}" for the more more prudent-minded) to facilitate package updates.

**Review by ambroisie — CHANGES_REQUESTED** [review]

**Comment by appsforartists** [comment]
I just rebuilt my flake with the suggested changes, and it seems to be working.

**Review by lukegb — APPROVED** [review]

**Pull request merged**
